### ISSUE TITLE
support for tuya diffuser

### ISSUE THREAD

**Issue by accelle17** [rospogrigio/localtuya#172]
Hello again,
Just wanted to ask if it's possible to add a custom component particularly for diffusers. I got below DPS and fan component doesn't work since the value for the speed is (off, small, big). 
The color effect needs to be set to particular effect before color or brightness can be set.
dps|current value|description|possible values
-- | -- | -- | --
'1' |true|master switch|true,false
'11'|true|light|true,false
'12'|0
'13'|'0'|diffuser countdown| on (1) off (0)
'14'|0|diffuser countdown timer|60, 180, 360
'103'|'small'|diffuser speed|off, small, big
'108'|'80000201666432'|hsv for dps:110 when set to 2
'109'|'white'
'110'|'2'|light effect|1: colorloop(gradient), 2: solid(fixed), 3: nightlight(white only)
'111'|255|brightness| 0-255 for dps:110 when set to 3 (nightlight)

I was thinking of the following settings at first:
1. switch: dps 1
2. light: dps 11, brightness-111,color mode-109,color-108,EFFECT-110
3. fan: dps 103

To Dos:
1. Add effect on light component
2. Add custom values on fan speed

**Comment by postlund** [comment]
Maybe @ThomasADavis can add the missing fan speeds and @ultratoto14 make the remaining changes to `light`?

Since we have DP 111 here, we should probably extend our DP detection for type d devices. Maybe something for you, @rospogrigio?

**Comment by ultratoto14** [comment]
Hi @postlund, I think that this device db may be needed in a very near future to support the increasing number of scenes.

@accelle17 so 111 is changing when 109 is in white and you change brightness ? What is the value of 109 when you change to light effect ? From you sentence, you can change the brightness of a scene ?

**Comment by postlund** [comment]
@ultratoto14 I agree and I will start working on it once I get my current PRs and issues under control. Shouldn't be long.

**Comment by accelle17** [comment]
> so 111 is changing when 109 is in white and you change brightness ? What is the value of 109 when you change to light effect ? From you sentence, you can change the brightness of a scene ?

Yes 111 is changing when 109 is white. It's usually at white unless 110 is change to 2 (solid, fixed) which is color.

**Comment by ultratoto14** [comment]
@postlund, i also have ideas, about that db, maybe we can open an enhancement issue where we can discuss about it ?

**Comment by postlund** [comment]
@ultratoto14 I'm working on just that, please let me finish that and we can continue from there.

**Comment by ultratoto14** [comment]
@accelle17 So it basically working as a usual tuya color light but the 110 need to also be forced to a specific value. 
It seems that 

1. Is the only real effect, colorloop
2. Is color
3. Is white

When you set up the light as you described, without setting up the scene, it's not changing at all ? No white to color, color to white, brightness ?

**Comment by accelle17** [comment]
yes, not changing to any, I can only do on/off. On tuya app too, you need to change the effect to fixed / nightlight in order to change color or brightness. 
Basically, you need to change to 110 to:

2: to make DPS 108 (HSV) working
3: to make DPS 111(brightness) working

**Comment by ultratoto14** [comment]
@accelle17 just to explain, as you are the first one to report this kind of behavior and this is a bit different from other bulbs, I need to know exactly how to introduce that, and how to also ease the configuration.

Here you have 3 settings, maybe others will have more or the same number but with different values

**Comment by postlund** [comment]
@ultratoto14 See #181.

**Comment by accelle17** [comment]
> @accelle17 just to explain, as you are the first one to report this kind of behavior and this is a bit different from other bulbs, I need to know exactly how to introduce that, and how to also ease the configuration.
> 
> Here you have 3 settings, maybe others will have more or the same number but with different values

Not sure if it's simple to just map DP 110 to light effects so that there's only two components to map:

1. Fan : dps 1, speed dps 103
2. light: dps 11, brightness:111,HSV-108,EFFECT:110

**Comment by ultratoto14** [comment]
@accelle17 currently effects color and white are mutually exclusive. As an example, if you set an effect to white, you will not be able to use brightness.

**Comment by accelle17** [comment]
> @accelle17 currently effects color and white are mutually exclusive. As an example, if you set an effect to white, you will not be able to use brightness.

Ok, might be better to ignore dps 111 and just stick to DPS 108 for whatever color/brightness as long as DPS 110 can be controlled.

**Comment by itairaz1** [comment]
Hi I have BW-FUN3
I'm not sure if I should post here or in #181 or open a new issue.
DPS mapping:
1: diffuser (spray) switch - True / False
2: diffuser fan - small / large (no "off" option)
3: diffuser countdown (in hours): 1 / 3 / 6 (no "off" or 0)
5: light switch - True / False
6: light mode - colour (by DPS 7) / colourful1 (switching colour. please note about "1" in the end)
7: light color - 0070ff00d5ffff

Note that diffuser fan (2) and countdown (3) will alway have a value, even when diffuser switch (1) is False

**Comment by ultratoto14** [comment]
@itairaz1 you cannot configure it to have all the features at the moment.
You may setup a fan with on/off and maybe the fan speed, and a light to have on/off and color management.

Is colourful1 a kind of scene that changes automatically the color ?

**Comment by itairaz1** [comment]
> @itairaz1 you cannot configure it to have all the features at the moment.
Thanks for reply!

> You may setup a fan with on/off and maybe the fan speed, and a light to have on/off and color management.
I did all the config already
I can see the current light color, but I can't change to specific one, when I'm trying to change the color it change to red
```
    entities:
      - platform: fan
        friendly_name: Diffuser
        id: 1
        fan_speed_control: 2 # doesnt work
      - platform: light
        friendly_name: Diffuser light
        id: 5 
        color_mode: 6
        color: 8
```
> Is colourful1 a kind of scene that changes automatically the color ?
It moving between colors in gradient style (sounds the same as colorloop)

**Comment by ultratoto14** [comment]
@itairaz1 can you set upper brightness to 255 for your light and see if it helps ?

**Comment by itairaz1** [comment]
@ultratoto14 didn't help

**Comment by itairaz1** [comment]
> 3: diffuser countdown (in hours): 1 / 3 / 6 (no "off" or 0)

update about that: 3 also can be "cancel" if countdown is disabled (unlimited time)

**Comment by ultratoto14** [comment]
@itairaz1 can you set some usual values for colors and note the value of the DPs ? I mean full Blue, Red and Green with brightness  at 30% and 100% ?

**Comment by itairaz1** [comment]
@ultratoto14 

using device button (each click):
ff00000000ffff - red
8000000000ff80 - dimmed red
ffff00003cffff - yellow
808000003cff80 - dimmed yellow
00ffff00b4ffff - blue
00808000b4ff80 - dimmed blue
00ff000078ffff - green
0080000078ff80 - dimmed green
(there are more colors...)

using app (only color picker, no control over brightness)
f0ff00003fffff - random yellow
ff44000010ffff - random orange
ff8c000020ffff - another orange
00ff2d0082ffff - random green
ff00f8012dffff - purpule

**Comment by itairaz1** [comment]
@ultratoto14 any updates?

**Comment by ultratoto14** [comment]
@itairaz1 no, sorry, nothing new.
Can you switch to master version in hacs ?
Are you able to make some code changes to test ?

Also can you do a full dps dump ? (check the wiki)

**Comment by itairaz1** [comment]
I've just recheck if after upgrading to the latest version (3.2) and it works like a charm!
I/You can add it to the table
Thanks!

**Comment by itairaz1** [comment]
@ultratoto14 I already did those, and I mapped DPSes
here is my last config
```
      - platform: fan
        friendly_name: Diffuser
        id: 1
        fan_speed_control: 2
        fan_speed_low: 'small' #works only after add "small" to fan.py:35
        fan_speed_high: 'large' #works only after add "large" to fan.py:41
      - platform: light
        friendly_name: Diffuser light
        id: 5
        color_mode: 6
        color: 8
        brightness_lower: 0
        brightness_upper: 255
        scene: 6 # This adds effect, when choosing one of the effects, it starts "colorful" and you can stop it only by turn off and on the light
```

**Comment by itairaz1** [comment]
I added PR in order to support small large values.
I also need to it to wiki / devices page.
@ultratoto14 should I add a new diffuser section?

**Comment by ultratoto14** [comment]
@itairaz1 i merged your PR. And added in the wiki. Happy new year.

**Comment by ultratoto14** [comment]
@itairaz1 if you can give us the product key (in HA integration -> devices) and the firmware, i will then add them to the wiki.

**Comment by itairaz1** [comment]
@ultratoto14 Thank! Happy new year!
Product key: 6dwaaq5egthwitlb
need to add also those to the wiki
```
        brightness_lower: 0
        brightness_upper: 255
        scene: 6
```

**Comment by ultratoto14** [comment]
Hi @itairaz1 will put brightness settings but scene is more a workaround than a real setting. So will need more investigation.

**Comment by fabioxp7980** [comment]
Hi all, as many of you the BW-FUN3 is no more flashable with easy with tuya-convert or soldering, I'm trying localtuya way, till now I was able to obtain the id and the key but if I add the device with localtuya integration it returns an error, the internal lan ip is automatically get (but in Tuya app I see only external ip), device firmware in 3.1 and I hav not updates, is there a way to use it with localtuya? Sorry if I ask but I followed all things and I think I'm not so far from solution. Thank you.

**Comment by dexsta** [comment]
> 
> 
> @ultratoto14 Thank! Happy new year! Product key: 6dwaaq5egthwitlb Firmware: 3.3 need to add also those to the wiki
> 
> ```
>         brightness_lower: 0
>         brightness_upper: 255
>         scene: 6
> ```

Could you kindly share how you have this working within HA. I can turn ON/OFF but unable to control speed. 
I get 3 values for the speed with the settings given from the WIKI or has something changed ? Thanks.

**Comment by dexsta** [comment]
From Dev Tools

![dev](https://user-images.githubusercontent.com/41833996/136639613-6e87016c-e1a9-4d57-b7b7-faa407de89a9.png)
